includes/Core/Util/Entity_Factory.php

```diff
@@ -484,4 +491,73 @@ private static function prefix_title( $title, $prefix ) {
 			$title
 		);
 	}
+
+	/**
+	 * Gets the entity for the given URL, if available.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $url URL to determine the entity from.
+	 * @param Entity $entity The initial entity.
+	 * @return Entity The initial or new entity for the given URL.
+	 */
+	private static function maybe_convert_to_amp_entity( $url, $entity ) {
+		if ( is_null( $entity ) || ! defined( 'AMP__VERSION' ) ) {
+			return $entity;
+		}
+
+		$url_parts   = wp_parse_url( $url );
+		$current_url = $entity->get_url();
+
+		if ( ! empty( $url_parts['query'] ) ) {
+			$url_query_params = array();
+
+			wp_parse_str( $url_parts['query'], $url_query_params );
+
+			// check if the $url has amp query param.
+			if ( array_key_exists( 'amp', $url_query_params ) ) {
+				$new_url = add_query_arg( 'amp', '1', $current_url );
+				return self::create_updated_amp_entity( $new_url, $entity );
+			}
+		}
+
+		if ( ! empty( $url_parts['path'] ) ) {
+			// We need to correctly add trailing slash if the original url had trailing slash.
+			// That's the reason why we need to check for both version.
+			if ( '/amp' === substr( $url_parts['path'], -4 ) ) { // -strlen('/amp') is -4
+				$new_url = untrailingslashit( $url_parts['path'] ) . '/amp';
+				return self::create_updated_amp_entity( $new_url, $entity );
+			}
+
+			if ( '/amp/' === substr( $url_parts['path'], -5 ) ) { // -strlen('/amp/') is -5
+				$new_url = untrailingslashit( $url_parts['path'] ) . '/amp/';
+				return self::create_updated_amp_entity( $new_url, $entity );
+			}
+		}
+
+		return $entity;
+	}
+
+	/**
+	 * Creates updated AMP Entity..
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $new_url URL of the new entity.
+	 * @param Entity $entity The initial entity.
+	 * @return Entity The new entity.
+	 */
+	private static function create_updated_amp_entity( $new_url, $entity ) {
```

Comment:
Also, let's rename this method. If we have `maybe_convert_to_amp_entity` method, then this method should be named `convert_to_amp_entity`.